### PR TITLE
Added the ability to disable/enable services

### PR DESCRIPTION
**Pull request by crypticC0der**
Adds commands and testing for the disabling and enabling of services, this is done through microovn enable <service> and microovn disable <service>.

This is deemed useful as the inability to control what nodes important services, such as central, are running on can mean a user may have their central database on lower performance nodes without the ability to move them.

### CODE REVIEW

File: microovn/ovn/services.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/canonical/microcluster/state"
+	"github.com/canonical/microovn/microovn/database"
+)
+
+func DisableService(s *state.State, service string) error {
+	err := s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
+		exists, err := database.ServiceExists(ctx, tx, s.Name(), service)
```
Comment (fnordahl) [inline microovn/ovn/services.go:15]: The `service` string is supplied by the end user of the program, we need to do some sort of validation of it before passing it on into the innards of the program.

Who knows what the user could craft to have the database or `snapctl` do unexpected things with elevated privileges.

File: microovn/ovn/services.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		if exists != true {
+			return errors.New("No such service")
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	err = snapStop(service, true)
```
Comment (fnordahl) [inline microovn/ovn/services.go:28]: Some of our services, such as `chassis` and `switch` are simple and map directly to a single systemd service.

However we do also have more complex services, such as `central` which in reality map into multiple systemd services:
https://github.com/canonical/microovn/blob/a99570ddbcffe6acbd76698a424c78d3cca49773/microovn/ovn/join.go#L113-L126

This is a useful abstraction and we want to keep the components individually controllable at the lower level.

However, for the purpose of the `microovn enable` and `microovn disable` commands we need some sort of abstraction so that the user can do `microovn disable central` and have the system disable all the underlying systemd services that compose that higher level service.

File: microovn/api/endpoints.go [hunk 3]
```diff
@@ -16,7 +17,9 @@ var Server = rest.Server{
 	Resources: []rest.Resources{
 		{
 			PathPrefix: types.APIVersion,
-			Endpoints: []rest.Endpoint{servicesCmd,
```
Comment (fnordahl) [inline microovn/api/endpoints.go:19]: Making the existing `servicesCmd` function public and renaming the module is a logical unit of change in itself. Please put that into a separate commit so that it is easier to discern between reorganization/refactoring of existing code and new code during review.

**Review by fnordahl — CHANGES_REQUESTED** [review]
Hello, MJ, and thank you for your work on this!

The approach overall looks good to me and there are some comments in-line requesting some changes.

In addition we would like to see some documentation for this new feature.

I'll defer to @mkalcok for any additional comments on implementation.

File: microovn/ovn/services.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		err := database.DeleteService(ctx, tx, s.Name(), service)
+		return err
+	})
+
+	return err
+
+}
+
+func EnableService(s *state.State, service string) error {
+	err := s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
```
Comment (mkalcok) [inline microovn/ovn/services.go:43]: There's a function that might give you the same functionality as this transaction
https://github.com/canonical/microovn/blob/a99570ddbcffe6acbd76698a424c78d3cca49773/microovn/node/services.go#L41-L43

File: tests/test_helper/bats/services.bats [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    for container in $TEST_CONTAINERS; do
+        run lxc_exec "$container" "microovn disable switch"
+        assert_output "Service switch disabled"
+    done
+}
+
+disable_non_enabled_service() {
+    for container in $TEST_CONTAINERS; do
+        run lxc_exec "$container" "microovn disable switch"
+        assert_output "Error: command failed: No such service"
```
Comment (mkalcok) [inline tests/test_helper/bats/services.bats:54]: I would suggest improving output of the command in this case (also in `enable_already_enabled_service`). Instead of `No such service`, which isn't really descriptive, it could say `Service <service_name> already disabled`. I'd even avoid all the `Error` and `command failed` prefixes.

File: tests/test_helper/bats/services.bats [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    load ${ABS_TOP_TEST_DIRNAME}test_helper/microovn.bash
+    load ${ABS_TOP_TEST_DIRNAME}../.bats/bats-support/load.bash
+    load ${ABS_TOP_TEST_DIRNAME}../.bats/bats-assert/load.bash
+
+    # Ensure TEST_CONTAINERS is populated, otherwise the tests below will
+    # provide false positive results.
+    assert [ -n "$TEST_CONTAINERS" ]
+}
+
+services_register_test_functions() {
```
Comment (mkalcok) [inline tests/test_helper/bats/services.bats:16]: These tests seem to be heavily dependent on order of execution which is not the best practice. If forced order of operations is really necessary, it should be a single test. Otherwise each test should restore state of the system on teardown.

As it may be a bit hard to restore system state on failure, I'd perhaps suggest to go with single long test here that bails in case of any failure.

File: tests/test_helper/bats/services.bats [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    bats_test_function \
+        --description "Disable non enabled service" \
+        -- disable_non_enabled_service
+    bats_test_function \
+        --description "Enable disabled service" \
+        -- enable_disabled_service
+}
+
+enable_disabled_service() {
+    for container in $TEST_CONTAINERS; do
```
Comment (mkalcok) [inline tests/test_helper/bats/services.bats:38]: suggestion: I'd say that it's enough to run the operation on single node. Is there a value in executing it on every node?

File: tests/test_helper/setup_teardown/basic_cluster.bash [hunk 8]
```diff
@@ -4,7 +4,7 @@ setup_file() {
     load test_helper/microovn.bash
 
 
-    TEST_CONTAINERS=$(container_names "$BATS_TEST_FILENAME" 3)
+    TEST_CONTAINERS=$(container_names "$BATS_TEST_FILENAME" 1)
```
Comment (mkalcok) [inline tests/test_helper/setup_teardown/basic_cluster.bash:7]: Why was this changed? `basic_cluster.bats` should operate on 3 node cluster.

**Review by mkalcok — CHANGES_REQUESTED** [review]
The content overall looks good to me, thank you MJ. I left couple of suggestions inline that should be considered.

**Comment by crypticC0der** [comment]
I have addressed the suggestions by @fnordahl and will implement the ones submitted by martin

File: microovn/api/services/control.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+// /1.0/services/service endpoint.
+var ServiceControlCmd = rest.Endpoint{
+	Path:   "service/{service}",
+	Delete: rest.EndpointAction{Handler: disableService, AllowUntrusted: false, ProxyTarget: true},
+	Put:    rest.EndpointAction{Handler: enableService, AllowUntrusted: false, ProxyTarget: true},
+}
+
+func enableService(s *state.State, r *http.Request) response.Response {
```
[the remark this answers is not in the document]
Comment (crypticC0der) [reply microovn/api/services/control.go:22]: I don't see why we would need to validate the input twice, that sounds unnecessary

File: microovn/ovn/services.go
[the remark this answers is not in the document]
Comment (crypticC0der) [reply microovn/ovn/services.go:1]: it uses the snap start and snap stop functions

File: microovn/api/services/control.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (mkalcok) [reply microovn/api/services/control.go:22]: If we manage to have single source of truth for these checks I don't think it's necessarily bad to validate input in multiple places. I think it makes sense to:
* Validate in request handler to stop processing invalid request as soon as possible
* Validate in function that actually implements the enable/disable functionality because there might be multiple code paths leading to that function call.

Though if you decide to go with one validation, I agree that the second one is more important. Just make sure that the invalid can not be misused for something else in the request handler.

File: microovn/cmd/microovn/disable.go
[the remark this answers is not in the document]
Comment (crypticC0der) [reply microovn/cmd/microovn/disable.go:1]: yes I agree

File: microovn/cmd/microovn/disable.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+type cmdDisable struct {
+	common  *CmdControl
+	cluster *cmdCluster
+}
+
+func (c *cmdDisable) Command() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "disable <SERVICE>",
+		Short: "disables a service",
```
[the remark this answers is not in the document]
Comment (crypticC0der) [reply microovn/cmd/microovn/disable.go:21]: to be discussed tomorrow

File: microovn/ovn/services.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+const (
+	SrvChassis SrvName = "chassis"
+	SrvCentral SrvName = "central"
+	SrvSwitch  SrvName = "switch"
+)
+var ServiceNames = []SrvName{SrvChassis,SrvCentral,SrvSwitch}
+
+func DisableService(s *state.State, service string) error {
+	err := s.Database.Transaction(s.Context, func(ctx context.Context, tx *sql.Tx) error {
```
Comment (fnordahl) [inline microovn/ovn/services.go:24]: We need to do validation of the `service` string prior to this call.

File: microovn/ovn/services.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+		return err
+	}
+
+	if SrvName(service) == SrvCentral{
+		err = snapStop("ovn-ovsdb-server-nb",true)
+		if err != nil {return err }
+		err = snapStop("ovn-ovsdb-server-sb",true)
+		if err != nil {return err }
+		err = snapStop("ovn-northd",true)
+	}
```
Comment (fnordahl) [inline microovn/ovn/services.go:44]: Nice!

While there exists a `central` systemd service in addition to the above ones, it is kept as a transitional service to avoid breaking anyone's workflow (see commit 595567eb1d33016a30ef76cb98433250ae4358b0).

For our purposes we should avoid using it, so let's add a `else` branch for the case of `service == SrvCentral`.

File: microovn/ovn/services.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+		err := database.DeleteService(ctx, tx, s.Name(), service)
+		return err
+	})
+
+	return err
+
+}
+
+func EnableService(s *state.State, service string) error {
+	exists, err := node.HasServiceActive(s, service)
```
Comment (fnordahl) [inline microovn/ovn/services.go:61]: We need to do validation of the service string prior to this call.

File: microovn/cmd/microovn/disable.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		return err
+	}
+
+	cli, err := m.LocalClient()
+	if err != nil {
+		return err
+	}
+
+	targetService := args[0]
```
Comment (fnordahl) [inline microovn/cmd/microovn/disable.go:39]: This will crash if no arguments are passed, i.e. if the user enters `microovn disable` without actually passing a service name.

File: microovn/cmd/microovn/disable.go [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
```
Comment (fnordahl) [inline microovn/cmd/microovn/disable.go:75]: This will crash if no arguments are passed, i.e. if the user enters `microovn enable` without actually passing a service name.

**Review by fnordahl — DISMISSED** [review]
In addition to the in-line comments `gofmt -d ./..` has opinions on formatting of some of the code.

I took this for a spin and while I am able to enable the central services on a fourth node using the `microovn enable central` command, there is currently no handling of actually making those services a part of the OVSDB clusters.

I realize that we have not yet specced that out, so I wonder whether to address that in this PR or if we should merge as-is and tackle that as the next challenge in the coming pulse. Thoughts?

File: tests/test_helper/bats/services.bats [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        run lxc_exec "$container" "microovn enable switchh"
+        assert_output "Error: Failed to enable service 'switchh': 'Service does not exist'"
+
+        # disable non existing service
+        run lxc_exec "$container" "microovn disable switchh"
+        assert_output "Error: Failed to disable service 'switchh': 'This service is not enabled'"
+
+        # disable service
+        run lxc_exec "$container" "microovn disable switch"
+        assert_output "Service switch disabled"
```
Comment (mkalcok) [inline tests/test_helper/bats/services.bats:38]: I forgot to mention this in my last review, but I think that the test should verify more than just command output. E.g:
* Check `microovn status` output to make sure that it reports expected services on the node
* Check `snap services microovn` to make sure that expected services were actually started

**Comment by mkalcok** [comment]
> I realize that we have not yet specced that out, so I wonder whether to address that in this PR or if we should merge as-is and tackle that as the next challenge in the coming pulse. Thoughts?

We could limit this feature to `switch` and `chassis` services for now, and add support for `central` in next pulse.

File: microovn/ovn/services.go (hunk 11 above)
Comment (crypticC0der) [reply microovn/ovn/services.go:24]: why, the service name gets validated within this call by checking it against the database?

Comment (fnordahl) [reply microovn/ovn/services.go:24]: It is a precaution in the event of vulnerable code in our library dependencies.

Our backing database is SQL-based, and history is riddled with security incidents rooted in SQL injection attacks.

While our code has solid mitigations for that by making use of [prepared statements](https://go.dev/doc/database/prepared-statements) with place holders for variables:
https://github.com/canonical/microovn/blob/a99570ddbcffe6acbd76698a424c78d3cca49773/microovn/database/service.mapper.go#L51-L55

which should avoid such issues, I'd rather be safe than sorry by validating the data before exposing it to our data store.

Comment (crypticC0der) [reply microovn/ovn/services.go:24]: I did look and check that the input is sanitized for the database queries, I shall validate the input in the API, though I stand by my opinion on it being somewhat unneccicary

**Comment by fnordahl** [comment]
> we have support for central, it can be enabled, which enables its child services due to a snap hook, and disabled which disables its child services directly

Indeed, but when you enable the central services on a fresh node for the first time. Does the clustered databases become a member of the cluster?

**Comment by crypticC0der** [comment]
> > we have support for central, it can be enabled, which enables its child services due to a snap hook, and disabled which disables its child services directly
> 
> Indeed, but when you enable the central services on a fresh node for the first time. Does the clustered databases become a member of the cluster?

okay thats a good point, I am unsure,

**Comment by mkalcok** [comment]
> it can be enabled, which enables its child services due to a snap hook

I'd really discourage the usage of  the `central` snap service. I think we should remove it for the `24.03` release as it's just a source of confusion.

**Comment by crypticC0der** [comment]
all that is left on this is
documentation, a list command for the services, a decision on what to do regarding central seemingly
all other issues have been fixed, and more tests have been added

**Comment by crypticC0der** [comment]
I have sorted out formatting now also

**Review by fnordahl — COMMENTED** [review]
From the perspective of enabling/disabling services functionality, this looks good to me. If we can have a commit documenting the new commands we should be ready.

For the `central` service there is additional work required to handle OVSDB server join and departure events which we can do as a follow-up if you'd like.

File: tests/test_helper/bats/services.bats [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+        run lxc_exec "$container" "snap services microovn | grep switch | grep enabled"
+        assert_output ""
+
+        # enable disabled service
+        run lxc_exec "$container" "microovn enable switch"
+        assert_output "Service switch enabled"
+
+        assert [ -n "$(run lxc_exec "$container" "microovn status | grep switch")"]
+        assert [ -n "$(run lxc_exec "$container" "snap services microovn | grep switch | grep enabled")"]
```
Comment (mkalcok) [inline tests/test_helper/bats/services.bats:55]: Couple of notes here:

* We could use the same sets of checks after service is disabled (just with the logic inverted)
* Wouldn't `microovn status | grep switch` produce false positives if the `switch` service is enabled on a different node?
* nit: `run` is a [function from BATS](https://bats-core.readthedocs.io/en/stable/writing-tests.html#run-test-other-commands) that stores output of the command in `$output` and return code in `$status`. More idiomatic way to approach this would be to use `run` in one statement and then do assert on the `$output`, or since `grep` will fail if it does not find a match, you could `run` and then [`assert_success`](https://github.com/bats-core/bats-assert?tab=readme-ov-file#assert_success)

**Review by mkalcok — DISMISSED** [review]
I have last few suggestions that I left in-line. Also, we don't have this stated outright in our `CONTRIBUTING.md`, but I think we should keep documenting code in [Go Doc style](https://go.dev/doc/comment), at least for functions and types. (wdyt @fnordahl ?)

File: tests/test_helper/bats/services.bats (hunk 17 above)
Comment (crypticC0der) [reply tests/test_helper/bats/services.bats:55]: there is only one node in this test

**Review by mkalcok — APPROVED** [review]
LGTM, thank you.

**Review by mkalcok — APPROVED** [review]
